src/builder/command.rs

```diff
@@ -4109,15 +4109,23 @@ impl<'help> Command<'help> {
 
         if !self.is_set(AppSettings::DisableHelpSubcommand) {
             debug!("Command::_check_help_and_version: Building help subcommand");
+            let help_about = "Print this message or the help of the given subcommand(s)";
             let mut help_subcmd = Command::new("help")
-                .about("Print this message or the help of the given subcommand(s)")
-                .arg(
-                    Arg::new("subcommand")
-                        .action(ArgAction::Append)
-                        .num_args(..)
-                        .value_name("SUBCOMMAND")
-                        .help("The subcommand whose help message to display"),
-                );
+                .about(help_about)
+                // help can take all other subcommands, including help, as a subcommand
+                .subcommands(
+                    self.get_subcommands()
```

Comment:
Doesn't this need to be recursive?

Reply:
Oh shoot, good catch, I completely missed that use case of `myprog help child grandchild` from https://github.com/clap-rs/clap/commit/2c12757bbdf34ce481f3446c074e24c09c2e60fd  
I was thinking `myprog child help grandchild` only.

That would definitely introduce a performance concern with the recursive subcommand tree copy.

(This current approach had the side benefit of adding a "SUBCOMMANDS" list when running `myprog help help`, but that is minor and not worth the performance hit.)

Sounds like a different approach is needed. Probably starting from adding a function to `clap_complete/src/generator/utils.rs` that copies the tree under help, and making sure each of the shells uses the function in its `Generator` impl.

Thanks for the review. I'll close this PR.

Reply:
Let's explore options first
- this all lives in `clap_complete`
- we have degrees of building a command (parse and complete)
- We leverage the difference between `build` and `_build_subcommand` to make a de facto degrees of building

Reply:
In other words, you're saying the parser could have a mode flag, telling it to build (slowly, ignoring performance) a full representation of the command tree that can be consumed elsewhere? It would be consumed by `clap_complete` as the shells need exactly this functionality. But also the tree, once built, could potentially serve as a cache to speed up future executions of the parser -- that would be why this logic could live in the parser?

Reply:
So my first implementation put `help` completion in the parser for these reasons:

* I thought it would be the easiest implementation, touching where the "help" subcommand is being created
* Would result in consistent `myprog help help` output with SUBCOMMANDS, matching the shell completion
* I did not consider the performance implications, since I missed the case that required recursion
* I was not yet aware of another central place to put the new code that all the shells could access (`clap_complete/src/generator/utils.rs`)

My previous comment was guessing at a reason to keep the code in the parser. I think we can disregard it as too hypothetical and outside the scope of help completion.

For now assuming I'll make another attempt, it would live fully in the `clap_complete` crate:

* Would avoid having to add more settings/flags to toggle parser logic
* Move slowness to the completion workflow, where performance doesn't matter as much
* I'm not currently aware of specific use cases where it needs to be more broadly available, to the parser or others. It's okay not to propagate to `help help` and `mangen` cases.

Reply:
Different areas of interest include
- completions: we want it all
- man pages: a user recursively generating them won't want this
- help output: this would provide a nicer view 
- usage: it would provide a nicer view

You are right, it probably will be fine for this to live in completions.  The downside is we'll need to reimplement this for the new completion work we are doing (`clap_complete/src/dynamic.rs`)